### PR TITLE
Non azure public cloud endpoint

### PR DESCRIPTION
**Pull request by theok-nice**
Addresses issue : [wal-g/wal-g/issues/#856](https://github.com/wal-g/wal-g/issues/856) 

Changes allow module to work in regions other than AzurePublicCloud. 
It adds optional config setting where user can select AzureUSGovernmentCloud, AzureChinaCloud, AzureGermanCloud.
When one of these three options is selected, module will select correct storage account endpoint. For the correct  endpoint I am using https://pkg.go.dev/github.com/Azure/go-autorest/autorest/azure module.

If setting is missing from the wal-g config file,or set to any other value (including "AzurePublicCloud") will default to AzurePublicCloud's endpoint.

### CODE REVIEW

File: azure/folder.go [hunk 1]
```diff
@@ -93,14 +99,26 @@ func ConfigureFolder(prefix string, settings map[string]string) (storage.Folder,
...
+	var storageEndpointSuffix string
+	switch environmentName {
+	case "AzureUSGovernmentCloud":
+		storageEndpointSuffix = azure.USGovernmentCloud.StorageEndpointSuffix
+	case "AzureChinaCloud":
+		storageEndpointSuffix = azure.ChinaCloud.StorageEndpointSuffix
+	case "AzureGermanCloud":
+		storageEndpointSuffix = azure.GermanCloud.StorageEndpointSuffix
+	default:
+		storageEndpointSuffix = azure.PublicCloud.StorageEndpointSuffix
```
Comment (na-jakobs) [inline azure/folder.go]: Indent this maybe?

Comment (theok-nice) [reply azure/folder.go]: Thank you for the comment  @na-jakobs

I am new in Golang and I am not sure what is the recommended way to indent switch statements, but most examples I have seen online follow the same logic.

For example
https://github.com/golang/go/wiki/Switch

**Comment by theok-nice** [comment]
@x4m or @usernamedt I am not sure who should review the MR in this repo. I just saw you merge most recently and thought to ping you. I hope its ok.

File: azure/folder.go [hunk 2]
```diff
@@ -65,11 +68,14 @@ func ConfigureFolder(prefix string, settings map[string]string) (storage.Folder,
 			return nil, NewCredentialError(AccessKeySetting)
 		}
 	}
+	if environmentName, ok = settings[EnvironmentName]; !ok {
+		environmentName = "AzurePublicCloud"
```
Comment (usernamedt) [inline azure/folder.go]: I would advise extracting this to a const at the top of the file and naming it like default environment

File: azure/folder.go [hunk 3]
```diff
@@ -93,14 +99,26 @@ func ConfigureFolder(prefix string, settings map[string]string) (storage.Folder,
 		return nil, NewFolderError(err, "Unable to create container")
 	}
 
+	var storageEndpointSuffix string
+	switch environmentName {
+	case "AzureUSGovernmentCloud":
```
Comment (usernamedt) [inline azure/folder.go]: same here, maybe also extract the long switch case to the separate method to improve readability

File: azure/folder.go (hunk 2 above)
Comment (theok-nice) [reply azure/folder.go]: Thank you for the advice. I created new const and set the default value "AzurePublicCloud" in my most recent [commit](https://github.com/wal-g/storages/pull/43/commits/ed5d7c1f42f0093181f1d3c30f0fe0c7849b937a)

File: azure/folder.go (hunk 3 above)
Comment (theok-nice) [reply azure/folder.go]: Thank you for the advice @usernamedt
I am not sure which variable you propose to have as const in your comment. In the meantime I moved switch case to a new function in my most recent [commit](https://github.com/wal-g/storages/pull/43/commits/ed5d7c1f42f0093181f1d3c30f0fe0c7849b937a).

File: azure/folder.go [hunk 4]
```diff
@@ -254,3 +245,21 @@ func getUploadStreamToBlockBlobOptions(settings map[string]string) azblob.Upload
...
 	return azblob.UploadStreamToBlockBlobOptions{MaxBuffers: maxBuffers, BufferSize: bufferSize}
 }
+
+// Function will get environment's name and return string with the environment's Azure storage account endpoint suffix.
+// Expected names AzureUSGovernmentCloud, AzureChinaCloud, AzureGermanCloud. If any other name is used the func will return
+// the Azure storage account endpoint suffix for AzurePublicCloud.
+func getStorageEndpointSuffix(environmentName string) string {
+	var storageEndpointSuffix string
+	switch environmentName {
+	case "AzureUSGovernmentCloud":
```
Comment (usernamedt) [inline azure/folder.go]: Replace "AzureUSGovernmentCloud" with azure.USGovernmentCloud.Name?

File: azure/folder.go [hunk 5]
```diff
@@ -254,3 +245,21 @@ func getUploadStreamToBlockBlobOptions(settings map[string]string) azblob.Upload
...
+		storageEndpointSuffix = azure.USGovernmentCloud.StorageEndpointSuffix
+	case "AzureChinaCloud":
```
Comment (usernamedt) [inline azure/folder.go]: same here

File: azure/folder.go [hunk 6]
```diff
@@ -254,3 +245,21 @@ func getUploadStreamToBlockBlobOptions(settings map[string]string) azblob.Upload
...
+		storageEndpointSuffix = azure.ChinaCloud.StorageEndpointSuffix
+	case "AzureGermanCloud":
```
Comment (usernamedt) [inline azure/folder.go]: same here

File: azure/folder.go (hunk 4 above)
Comment (theok-nice) [reply azure/folder.go]: For each case name, I used the environment name that Microsoft is returning when use [Azure Instance Metadata Service](https://docs.microsoft.com/en-us/azure/virtual-machines/linux/instance-metadata-service?tabs=linux). Also they are the same names as the ones used in the variables of the [module](https://pkg.go.dev/github.com/Azure/go-autorest/autorest/azure#pkg-variables) that I am using.

The plan is to have the user that is configuring wal-g, to set/select the environment name that want to use. 
I believe the names I have are simpler to use vs `azure.USGovernmentCloud.Name`

File: azure/folder.go (hunk 5 above)
Comment (theok-nice) [reply azure/folder.go]: See answer in [previous comment](https://github.com/wal-g/storages/pull/43#discussion_r569373883).

File: azure/folder.go (hunk 6 above)
Comment (theok-nice) [reply azure/folder.go]: See answer in [previous comment](https://github.com/wal-g/storages/pull/43#discussion_r569373883).

File: azure/folder.go (hunk 4 above)
Comment (usernamedt) [reply azure/folder.go]: It looks like they are the same:
https://github.com/Azure/go-autorest/blob/c0fe8b0fea3d7344cc0a3e332c1bf28006c7fd83/autorest/azure/environments.go#L84

https://github.com/Azure/go-autorest/blob/c0fe8b0fea3d7344cc0a3e332c1bf28006c7fd83/autorest/azure/environments.go#L121

and so on

My intention was to get rid of hardcoded values by replacing them with library defined ones.

Comment (theok-nice) [reply azure/folder.go]: Took me a few mins to realize but in the end I understand what you mean. I made the [change](https://github.com/wal-g/storages/pull/43/commits/ea752ec5f3f8335cd58a7d179da4d841c3337730 )

File: azure/folder.go (hunk 5 above)
Comment (theok-nice) [reply azure/folder.go]: Took me a few mins to realize but in the end I understand what you mean. I made the change

File: azure/folder.go (hunk 6 above)
Comment (theok-nice) [reply azure/folder.go]: fixed

**Review by usernamedt — APPROVED** [review]

**Comment by theok-nice** [comment]
@usernamedt thank you for your time. 
I have one more question: now that the new code was merge to master, how is the module updated in 
https://pkg.go.dev/github.com/wal-g/storages  ?

Its the first I do something like this in go, and I can't find a straight forward answer on internet.

**Comment by usernamedt** [comment]
> @usernamedt thank you for your time. 
> 
> I have one more question: now that the new code was merge to master, how is the module updated in 
> 
> https://pkg.go.dev/github.com/wal-g/storages  ?
> 
> 
> 
> Its the first I do something like this in go, and I can't find a straight forward answer on internet.

Hi! Thanks for making wal-g better :)

It usually takes some time to update the version’s cache in pkg.go.dev.

Also, please note that you need an additional PR to wal-g/wal-g in order to add the new setting to the wal-g allowed settings list.

Also, you need to bump wal-g/storages version in wal-g/wal-g repo.

**Comment by theok-nice** [comment]
> Hi! Thanks for making wal-g better :)
> 
> It usually takes some time to update the version’s cache in pkg.go.dev.
> 
> Also, please note that you need an additional PR to wal-g/wal-g in order to add the new setting to the wal-g allowed settings list.
> 
> Also, you need to bump wal-g/storages version in wal-g/wal-g repo.

Glad that I can contribute on the project. 

I wasn't sure if the pkg.go.dev will update automatically, or if there was some other procedure, that's why I am asking. 

Yes my next step is create the PR in wal-g/wal-g. I am just waiting for the changes to show up in pkg.go.dev first. 

Thanks again for your time.

**Pull request merged**